README.rst
Celery One
===========

|Build Status|

Celery One allows you to prevent multiple execution and queuing of `celery <http://www.celeryproject.org/>`_ tasks.
The project is a fork of `celery-once <https://github.com/TrackMaven/celery-once>`_ to which I give all credits
for the initial architecture and part of this README :). The project was created to add more functionality to the original version.
``celery_one`` is compatible with ``celery_once``'s previous behaviour if properly configured.

The main differences are the following:

* Option to prevent tasks with same id
* Option to get an AsyncResult if the task is already queue
* Compatibility with chords and groups
* Optional infinite timeout
* Redis connection pool

Installation
============

To install ``celery_one`` with pip just run:

::

     pip install -e git+https://github.com/szandara/celery-one.git#egg=celery_one


Requirements
============

* `Celery <http://www.celeryproject.org/>`_. Built to run with Celery 3.1. Older versions may work, but are not officially supported.
* `Redis <http://redis.io/>`_ is used as a distributed locking mechanism.

Simple Usage
=====

To use ``celery_one``, your tasks need to inherit from an `abstract <http://celery.readthedocs.org/en/latest/userguide/tasks.html#abstract-classes>`_ base task called ``QueueOne``.

You may need to tune the following Celery configuration options...

    * ``ONE_REDIS_URL`` should point towards a running Redis instance (defaults to ``redis://localhost:6379/0``)
    * ``ONE_DEFAULT_TIMEOUT`` how many seconds after a lock has been set before it should automatically timeout (defaults to 3600 seconds, or 1 hour).


.. code:: python

    from celery import Celery
    from celery_one import QueueOne
    from time import sleep

    celery = Celery('tasks', broker='amqp://guest@localhost//')
    celery.conf.ONE_REDIS_URL = 'redis://localhost:6379/0'
    celery.conf.ONE_DEFAULT_TIMEOUT = 60 * 60

    @celery.task(base=QueueOne)
    def slow_task():
        sleep(2)
        return "Done!"


Behind the scenes, this overrides ``apply_async`` and ``delay``. To do so, ``celery_one`` uses Redis. Each time a task
the user calls an asynchronous task, it locks a key into Redis. If a key with a same key is pushed to the queue, its
execution is prevented by either raising an ``AlreadyQueued`` exception or returning the AsyncResult of the already running task.
The key of the task is generated from its name and parameters.

.. code:: python

    example.delay(10)
    example.delay(10)
    Traceback (most recent call last):
        ..
    AlreadyQueued()

.. code:: python

    result1 = example.apply_async(args=(10,))
    result2 = example.apply_async(args=(10,), one_options={'fail':False}) # No failure here
    assert result1.id == result2.id

    result = example.apply_async(args=(10,))
    Traceback (most recent call last):
        ..
    AlreadyQueued()

Options
=====

``use_id``
------------

If this option is set, the key will be generated using the id of the task. This can be useful when working
with meaningful task ids or when the arguments are not necessarily indication of different tasks.

.. code:: python

    @celery.task(base=QueueOne, one_options={'use_id':True})
    def slow_task_no_fail(a):
        print("Running")
        sleep(10)
        return "Done: " + str(a)

    result1 = slow_task_no_fail.apply_async(args=(10), task_id=id1)
    result2 = slow_task_no_fail.apply_async(args=(12), task_id=id1)

    Traceback (most recent call last):
        ..
    AlreadyQueued()


``fail``
------------

Optionally, instead of raising an ``AlreadyQueued`` exception, the task can return an `AsyncResult <http://docs.celeryproject.org/en/latest/reference/celery.result.html>`_.
To do so, set the option in the celery task or directly in the ``apply_async`` call.

.. code:: python

    @celery.task(base=QueueOne, one_options={'fail':False})
    def slow_task_no_fail():
        print("Running")
        sleep(2)
        return "Done!"

    result1 = slow_task_no_fail.apply_async(args=(10))
    result2 = slow_task_no_fail.apply_async(args=(10))

    print(result1.get())
    print(result2.get())

    Output:

    Running
    Done!
    Done!


``keys``
--------

By default ``QueueOne`` creates a lock based on the task's name and its arguments and values.
Take for example, the following task below...

.. code:: python

    @celery.task(base=QueueOne)
    def slow_add(a, b):
        sleep(2)
        return a + b

Running the task with different arguments will default to checking against different locks.

.. code:: python

    slow_add(1, 1)
    slow_add(1, 2)

If you want to specify locking based on a subset, or no arguments you can adjust the keys ``celery_one`` looks at in the task's `options <http://celery.readthedocs.org/en/latest/userguide/tasks.html#list-of-options>`_ with ``one_options={'keys': [..]}``

.. code:: python

    @celery.task(base=QueueOne, one_options={'keys': ['a']})
    def slow_add(a, b):
        sleep(30)
        return a + b

    example.delay(1, 1)
    # Checks if any tasks are running with the `a=1`
    example.delay(1, 2)
    Traceback (most recent call last):
        ..
    AlreadyQueued()
    example.delay(2, 2)

.. code:: python

    @celery.task(base=QueueOne, one_options={'keys': []})
    def slow_add(a, b):
        sleep(30)
        return a + b

    # Will enforce only one task can run, no matter what arguments.
    example.delay(1, 1)
    example.delay(2, 2)
    Traceback (most recent call last):
        ..
    AlreadyQueued()


``timeout``
-----------
As a fall back, ``celery_one`` will clear a lock after 60 minutes.
This is set globally in Celery's configuration with ``ONE_DEFAULT_TIMEOUT`` but can be set for individual tasks using...

.. code:: python

    @celery.task(base=QueueOne, one_options={'timeout': 60 * 60 * 10})
    def long_running_task():
        sleep(60 * 60 * 3)

``timeout`` can also be set to None, causing ``celery_one`` to prevent adding same tasks until the already running one
is complete. *NOTE*: This might result in dangerous behaviors such as deadlocks or failing task executions. Use with care!

``unlock_before_run``
---------------------

By default, the lock is removed after the task has executed (using celery's `after_return <https://celery.readthedocs.org/en/latest/reference/celery.app.task.html#celery.app.task.Task.after_return>`_). This behaviour can be changed setting the task's option ``unlock_before_run``. When set to ``True``, the lock will be removed just before executing the task.

**Caveat**: any retry of the task won't re-enable the lock!

.. code:: python

    @celery.task(base=QueueOne, one_options={'unlock_before_run': True})
    def slow_task():
        sleep(30)
        return "Done!"


Support
=======

* Tests are run against Python 2.7 and 3.3. Other versions may work, but are not officially supported.

Contributing
============

Contributions are welcome, and they are greatly appreciated! See `contributing
guide <CONTRIBUTING.rst>`_ for more details.


.. |Build Status| image:: https://travis-ci.org/szandara/celery-one.svg
   :target: https://travis-ci.org/szandara/celery-one

HISTORY.rst
History
=======

0.1
---

-  Initial fork from celery-once

tests/conftest.py
from fakeredis import FakeStrictRedis

import pytest


@pytest.fixture()
def redis(monkeypatch):
    fake_redis = FakeStrictRedis()
    fake_redis.flushall()
    monkeypatch.setattr("celeryone.celeryone.QueueOne.redis", fake_redis)
    return fake_redis


tests/unit/test_task.py
import mock
from celery import task
from celeryone import QueueOne, AlreadyQueued
from freezegun import freeze_time
import pytest


@task(name='simple_example', base=QueueOne)
def simple_example():
    return "simple"


@task(name='bound_task', bind=True, base=QueueOne)
def bound_task(self, a, b):
    return a + b


@task(name='args_example', base=QueueOne)
def args_example(a, b):
    return a + b


@task(name='select_args_example', base=QueueOne, one_options={'keys': ['a']})
def select_args_example(a, b):
    return a + b


def test_get_key_simple():
    assert "qo_simple_example" == simple_example.get_key_from_args()


def test_get_key_args_1():
    assert "qo_args_example_a-1_b-2" == args_example.get_key_from_args(kwargs={'a':1, 'b': 2})


def test_get_key_args_2():
    assert "qo_args_example_a-1_b-2" == args_example.get_key_from_args(args=(1, 2, ))


def test_get_key_select_args_1():
    assert "qo_select_args_example_a-1" == select_args_example.get_key_from_args(kwargs={'a':1, 'b': 2})


def test_get_key_bound_task():
    assert "qo_bound_task_a-1_b-2" == bound_task.get_key_from_args(kwargs={'a': 1, 'b': 2})


@freeze_time("2012-01-14")  # 1326499200
def test_raise_or_lock(redis):
    print(redis.hgetall("test"))
    assert not bool(redis.hgetall("test"))
    QueueOne().raise_or_lock(key="test", queue_one_data={'task_id': 'test_id', 'expires': 60})
    assert bool(redis.hgetall("test"))
    assert redis.ttl("test") == 60


@freeze_time("2012-01-14")  # 1326499200
def test_raise_or_lock_locked(redis):
    # Set to expire in 30 seconds!
    redis.hmset("test", {'task_id': 'test_id', 'expires': 1326499200 + 30})
    with pytest.raises(AlreadyQueued) as e:
        QueueOne().raise_or_lock(key="test", queue_one_data={'task_id': 'test_id', 'expires': 60})
    assert e.value.countdown == 30
    assert e.value.message == "Task test_id expires in 30 seconds"

@freeze_time("2012-01-14")  # 1326499200
def test_raise_or_lock_locked_and_expired(redis):
    # Set to have expired 30 ago seconds!
    redis.hmset("test", {'task_id': 'test_id', 'expires': 1326499200 - 30})
    QueueOne().raise_or_lock(key="test", queue_one_data={'task_id': 'test_id', 'expires': 60})
    assert bool(redis.hgetall("test"))
    assert redis.ttl("test") == 60

def test_clear_lock(redis):
    redis.hmset("test", {'task_id': 'test_id', 'expires': 1326499200 + 30})
    QueueOne().clear_lock("test")
    assert not bool(redis.hgetall("test"))


@task(name='task_id_example', base=QueueOne, one_options={'keys': ['a']})
def task_id_example(a, b):
    return a + b

def test_task_id_key():
    assert "qo_task_id_example_a_key" == task_id_example.get_key_from_id(task_id='a_key')

def test_task_id_key_num():
    assert "qo_task_id_example_10" == task_id_example.get_key_from_id(task_id=10)

@mock.patch('celeryone.celeryone.get_redis')
def test_redis_cached_property(get_redis_mock):
    # Remove any side effect previous tests could have had
    del simple_example.app._cached_redis

    assert get_redis_mock.call_count == 0
    assert not hasattr(simple_example.app, '_cached_redis')
    simple_example.redis
    assert get_redis_mock.call_count == 1
    assert hasattr(simple_example.app, '_cached_redis')
    simple_example.redis
    assert hasattr(simple_example.app, '_cached_redis')
    # as simple_example.redis is a cached_property, get_redis has not been called again
    assert get_redis_mock.call_count == 1


tests/unit/test_helpers.py
# -*- coding: utf-8 -*-
from celeryone.helpers import queue_one_key, kwargs_to_list, force_string

import pytest
import six


def test_force_string_1():
    assert force_string('a') == 'a'


def test_force_string_2():
    assert force_string(u'a') == 'a'


def test_force_string_3():
    assert force_string('é') == 'é'


def test_force_string_4():
    assert force_string(u'é') == 'é'


def test_kwargs_to_list_empty():
    keys = kwargs_to_list({})
    assert keys == []


def test_kwargs_to_list_1():
    keys = kwargs_to_list({'int': 1})
    assert keys == ["int-1"]


def test_kwargs_to_list_2():
    keys = kwargs_to_list({'int': 1, 'boolean': True})
    assert keys == ["boolean-True", "int-1"]


def test_kwargs_to_list_3():
    keys = kwargs_to_list({'int': 1, 'boolean': True, 'str': "abc"})
    assert keys == ["boolean-True", "int-1", "str-abc"]


def test_kwargs_to_list_4():
    keys = kwargs_to_list(
        {'int': 1, 'boolean': True, 'str': 'abc', 'list': [1, '2']})
    assert keys == ["boolean-True", "int-1", "list-[1, '2']", "str-abc"]


@pytest.mark.skipif(six.PY3, reason='requires python 2')
def test_kwargs_to_list_5():
    keys = kwargs_to_list(
        {'a': {u'é': 'c'}, 'b': [u'a', 'é'], u'c': 1, 'd': 'é', 'e': u'é'})
    assert keys == [
        "a-{'\\xc3\\xa9': 'c'}",
        "b-['a', '\\xc3\\xa9']",
        "c-1",
        "d-\xc3\xa9",
        "e-\xc3\xa9",
    ]


@pytest.mark.skipif(six.PY2, reason='requires python 3')
def test_kwargs_to_list_6():
    keys = kwargs_to_list(
        {'a': {u'é': 'c'}, 'b': [u'a', 'é'], u'c': 1, 'd': 'é', 'e': u'é'})
    assert keys == ["a-{'é': 'c'}", "b-['a', 'é']", "c-1", "d-é", 'e-é']


def test_queue_one_key():
    key = queue_one_key("example", {})
    assert key == "qo_example"


def test_queue_one_key_kwargs():
    key = queue_one_key("example", {'pk': 10})
    assert key == "qo_example_pk-10"


def test_queue_one_key_kwargs_restrict_keys():
    key = queue_one_key("example", {'pk': 10, 'id': 10}, restrict_to=['pk'])
    assert key == "qo_example_pk-10"


@pytest.mark.skipif(six.PY3, reason='requires python 2')
def test_queue_one_key_unicode_py2():
    key = queue_one_key(u"éxample", {'a': u'é', u'b': 'é'})
    assert key == "qo_\xc3\xa9xample_a-\xc3\xa9_b-\xc3\xa9"


@pytest.mark.skipif(six.PY2, reason='requires python 3')
def test_queue_one_key_unicode_py3():
    key = queue_one_key(u"éxample", {'a': u'é', u'b': 'é'})
    assert key == "qo_éxample_a-é_b-é"


celeryone/helpers.py
# -*- coding: utf-8 -*-
from time import time
from redis import StrictRedis
try:
    from urlparse import urlparse
except:
    # Python 3!
    from urllib.parse import urlparse
import six


def cached_celery_property(f):
    """
    Cache a property at the celery application level.
    """
    attr_name = '_cached_' + f.__name__

    @property
    def _cached_property(self):
        if not hasattr(self.app, attr_name):
                setattr(self.app, attr_name, f(self))
        return getattr(self.app, attr_name)
    return _cached_property


def parse_redis_details(url):
    parsed = urlparse(url)
    details = {
        'host': parsed.hostname,
        'password': parsed.password,
        'port': parsed.port
    }
    try:
        details['db'] = int(parsed.path.lstrip('/'))
    except:
        pass
    return details


def get_redis(url):
    return StrictRedis(**(parse_redis_details(url)))


def now_unix():
    """
    Returns the current time in UNIX time.
    """
    return int(time())


def force_string(kwargs):
    """
    Force key in dict or list to a string.
    Fixes: https://github.com/TrackMaven/celery-once/issues/11
    """
    if isinstance(kwargs, dict):
        return {
            force_string(key): force_string(value) for key, value in six.iteritems(kwargs)}
    elif isinstance(kwargs, list):
        return [force_string(element) for element in kwargs]
    elif six.PY2 and isinstance(kwargs, unicode):
        return kwargs.encode('utf-8')
    return kwargs


def kwargs_to_list(kwargs):
    """
    Turns {'a': 1, 'b': 2} into ["a-1", "b-2"]
    """
    kwargs_list = []
    # Kwargs are sorted in alphabetic order by their keys.
    # Taken from http://www.saltycrane.com/blog/2007/09/how-to-sort-python-dictionary-by-keys/
    for k, v in sorted(six.iteritems(kwargs), key=lambda kv: str(kv[0])):
        kwargs_list.append(str(k) + '-' + str(force_string(v)))
    return kwargs_list


def queue_one_key(name, kwargs, restrict_to=None):
    """
    Turns a list the name of the task, the kwargs and allowed keys
    into a redis key.
    """
    keys = ['qo', force_string(name)]
    # Restrict to only the keys allowed in keys.
    if restrict_to is not None:
        restrict_kwargs = {key: kwargs[key] for key in restrict_to}
        keys += kwargs_to_list(restrict_kwargs)
    else:
        keys += kwargs_to_list(kwargs)
    key = "_".join(keys)
    return key


celeryone/celeryone.py

from celery import Task, task
from celery.result import AsyncResult
from inspect import getcallargs

import base64
from .helpers import queue_one_key, get_redis, now_unix, cached_celery_property

class AlreadyQueued(Exception):
    def __init__(self, task_id, countdown):
        self.message = "Task {} expires in {} seconds".format(task_id, countdown)
        self.countdown = countdown
        self.task_id = task_id


class QueueOne(Task):
    AlreadyQueued = AlreadyQueued
    one_options = {
        'fail': False,
        'unlock_before_run': False,
        'use_id': False
    }

    """
    'There can be only one'. - Highlander (1986)

    An abstract tasks with the ability to detect if it has already been queued.
    When running the task (through .delay/.apply_async) it checks if the tasks
    is not already queued.
    """
    abstract = True
    one_options = {}

    @property
    def config(self):
        app = self._get_app()
        return app.conf

    @cached_celery_property
    def redis(self):
        """
        Return a connection pool to redis.

        This property is cached at the celery app level, so that only the first
        executed QueueOne task needs to instanciate a redis connection. The
        other ones will only re-use it.

        """
        return get_redis(
            getattr(self.config, "ONE_REDIS_URL", "redis://localhost:6379/0"))

    @property
    def default_timeout(self):
        return getattr(
            self.config, "ONE_DEFAULT_TIMEOUT", 60 * 60)

    def apply_async(self, args=None, kwargs=None, **options):
        """
        Queues a task using its task_id, raises an exception by default if already queued.
        The task_id must be set for this task type.

        :param \*args: positional arguments passed on to the task.
        :param \*\*kwargs: keyword arguments passed on to the task.
        :keyword \*\*one_options: (optional)
            :param: fail: (optional)
                If True, wouldn't raise an exception if already queued.
                Instead will return AsyncResult(task_id).
            :param: timeout: (optional)
                An `int' number of seconds after which the lock will expire.
                If not set, defaults to 1 hour.
            :param: use_id: (optional)
                If True, it will try to use the task_id to identify a same running task

        """

        # Has the user set the id?
        has_no_id = 'task_id' not in options

        task_id = options.get('task_id', None)

        if task_id:
            task_id = str(task_id)

        # Is this task part of a group?
        is_in_group = 'group_id' in options

        one_options = options.get('one_options', {})

        use_id = one_options.get('use_id', self.one_options.get('use_id', False))
        must_fail = one_options.get('fail', self.one_options.get('fail', True))
        timeout = one_options.get('timeout', self.one_options.get('expires', self.default_timeout))

        key = self.get_key(one_options, task_id, args, kwargs)

        if not use_id and not task_id:
            task_id = self.generate_id_from_key(key)

        if has_no_id:
            options['task_id'] = task_id
        celery_uno_data = {'task_id': task_id, 'expires': timeout}

        try:
            self.raise_or_lock(key, celery_uno_data)
        except self.AlreadyQueued as e:
            if not must_fail:
                if is_in_group:
                    return no_op.apply_async(**options)
                else:
                    return AsyncResult(e.task_id)
            raise e
        return super(QueueOne, self).apply_async(args, kwargs, **options)

    def generate_id_from_key(self, key):
        """
        Generate a b64 representation of the queue one key
        :param key:
        :return:
        """
        return 'qoid_' + base64.b64encode((self.name + key).encode('utf-8')).decode()

    def get_key(self, one_options, task_id, args, kwargs):

        use_id = one_options.get('use_id', self.one_options.get('use_id', False))
        # The user can specify whether he wants to use the task id or the parameters
        if use_id:
            if not task_id:
                raise ValueError('Asked to use the task_id but the task_id is not set')
            key = self.get_key_from_id(task_id)
        else:
            key = self.get_key_from_args(args, kwargs)

        return key

    def get_key_from_id(self, task_id=None):
        """
        Generate the key from the id of the task
        """
        if not task_id:
            raise ValueError('Task id cannot be null')

        keys = ['qo', self.name, str(task_id)]
        return '_'.join(keys)

    def get_key_from_args(self, args=None, kwargs=None):
        """
        Generate the key from the name of the task (e.g. 'tasks.example') and
        args/kwargs.
        """
        restrict_to = self.one_options.get('keys', None)
        args = args or {}
        kwargs = kwargs or {}
        call_args = getcallargs(self.run, *args, **kwargs)

        # Remove the task instance from the kwargs. This only happens when the
        # task has the 'bind' attribute set to True. We remove it, as the task
        # has a memory pointer in its repr, that will change between the task
        # caller and the celery worker
        if isinstance(call_args.get('self'), Task):
            del call_args['self']
        key = queue_one_key(self.name, call_args, restrict_to)
        return key

    def raise_or_lock(self, key, queue_one_data):
        """
        Checks if the task is locked and raises an exception, else locks
        the task.
        """
        now = now_unix()
        # Check if the tasks is already queued if key is in redis.

        running_task = self.redis.hgetall(key)

        if bool(running_task):
            running_task_id = running_task.get(b'task_id', None).decode()
            running_expire_time = running_task.get(b'expires', None)

            if running_expire_time:
                running_expire_time = running_expire_time.decode()
                remaining = int(running_expire_time) - now
                if remaining > 0:
                    raise self.AlreadyQueued(running_task_id, remaining)

            elif running_task_id:
                raise self.AlreadyQueued(running_task_id, None)

        expire_time = queue_one_data['expires']

        # If the user requires an expire time we set it manually on the hset
        if expire_time:
            timeout = queue_one_data['expires']
            queue_one_data['expires'] += now

            self.redis.hmset(key, queue_one_data)

            self.redis.expire(key, timeout)
        else:
            self.redis.hmset(key, queue_one_data)

    def get_unlock_before_run(self):
        return self.one_options.get('unlock_before_run', False)

    def clear_lock(self, key):
        self.redis.delete(key)

    def after_return(self, status, retval, task_id, args, kwargs, einfo):
        """
        After a task has run (both successfully or with a failure) clear the
        lock if "unlock_before_run" is False.
        """
        # Only clear the lock after the task's execution if the
        # "unlock_before_run" option is False
        if not self.get_unlock_before_run():
            key = self.get_key(self.one_options, task_id, args, kwargs)
            self.clear_lock(key)

    def __call__(self, *args, **kwargs):
        # Only clear the lock before the task's execution if the
        # "unlock_before_run" option is True

        if self.get_unlock_before_run():
            key = self.get_key(self.one_options, kwargs.get('task_id', None), args, kwargs)
            self.clear_lock(key)

        return super(QueueOne, self).__call__(*args, **kwargs)

@task(name='no_op')
def no_op():
    """
    This task is used to perform an empty operation. It is used to make celery uno compatible with
    groups.
    :return:
    """
    pass
